Frameworks/CoreText/DWriteWrapper_CTFont.mm

```diff
@@ -208,6 +208,207 @@ HRESULT _DWriteCreateFontFaceWithFontDescriptor(CTFontDescriptorRef fontDescript
+    } else if (CFStringFindWithOptions(key, kCTFontNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByInformationalString(fonts,
+                                                            static_cast<CFStringRef>(value),
+                                                            DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME,
+                                                            matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontDisplayNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByInformationalString(fonts,
+                                                            static_cast<CFStringRef>(value),
+                                                            DWRITE_INFORMATIONAL_STRING_FULL_NAME,
+                                                            matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontStyleNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByStyleName(fonts, static_cast<CFStringRef>(value), matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontTraitsAttribute, range, 0, nullptr)) {
+        __FilterFontsByTraits(fonts, static_cast<CFDictionaryRef>(value), matchingFonts);
+    } else {
+        // Other attributes are currently unsupported by CTFont
+        UNIMPLEMENTED();
+    }
+
+    // Do not filter if there's no match and forceMatch isn't true
+    // Used to share filtering between mandatory and optional attributes
+    if (!matchingFonts.empty() || forceMatch) {
+        fonts = std::move(matchingFonts);
+    }
+
+    return S_OK;
+}
+
+#pragma endregion // Helper Filter Methods
+
+/**
+ * Creates a CFArray of CTFontDescriptors values for all fonts that match the given attributes best and completely match mandatoryKeys
+ */
+HRESULT _DWriteCreateMatchingFontDescriptors(CFDictionaryRef attributes, CFSetRef mandatoryKeys, CFArrayRef* matchingNames) {
+    RETURN_HR_IF(E_POINTER, !attributes || !matchingNames);
+    *matchingNames = nullptr;
+
+    std::vector<ComPtr<IDWriteFont>> fonts;
+    if (CFDictionaryContainsKey(attributes, kCTFontNameAttribute) &&
+        (!mandatoryKeys || CFSetContainsValue(mandatoryKeys, kCTFontNameAttribute))) {
+        // Font name is either mandatory or there are no mandatory keys, so we can start match from font
+        ComPtr<IDWriteFont> matchingFont;
+        if (FAILED(_DWriteCreateFontWithName(static_cast<CFStringRef>(CFDictionaryGetValue(attributes, kCTFontNameAttribute)),
+                                             &matchingFont))) {
+            if (mandatoryKeys) {
+                // Not matching is not a fatal error, but we can just return now rather than try to match against any other attributes
+                return S_FALSE;
+            }
+        } else {
+            fonts.emplace_back(matchingFont);
+        }
+    }
+
+    if (fonts.empty() && CFDictionaryContainsKey(attributes, kCTFontFamilyNameAttribute) &&
```

Comment:
No, because font if we've gone into fontName and it failed but it's not required we still want to go into the family branch.